### PR TITLE
✨ Use new optimiser in internal modules

### PR DESCRIPTION
**Pull request by je-cook**
## Description

Switches over to new optimisation API internally for remaining modules

## Interface Changes

Aims to be none

## Checklist

I confirm that I have completed the following checks:

- [ ] Tests run locally and pass `pytest tests --reactor`
- [ ] Code quality checks run locally and pass `flake8` and `black .`
- [ ] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

File: bluemira/equilibria/optimisation/problem/_minimal_current.py [hunk 1]
```diff
@@ -41,12 +41,16 @@ class MinimalCurrentCOP(CoilsetOptimisationProblem):
...
+        Coilset to optimise
     eq:
         Equilibrium object to optimise the currents for
-    optimiser:
-        Optimiser object to use
     max_currents:
         Current bounds vector [A]
+    opt_conditions:
+        Optimiser conditions
+    opt_algorithm:
```
Comment (oliverfunk) [inline bluemira/equilibria/optimisation/problem/_minimal_current.py:52]: On line 64, the typing should be `Union[str, Algorithm] = "SLSQP",`

File: bluemira/equilibria/optimisation/problem/_nested_position.py [hunk 2]
```diff
@@ -98,10 +96,7 @@ def __init__(
         targets: MagneticConstraintSet,
         position_mapper: PositionMapper,
         opt_algorithm="SBPLX",
-        opt_conditions={
-            "stop_val": 1.0,
-            "max_eval": 100,
-        },
+        opt_conditions: Optional[Dict[str, float]] = None,
```
Comment (oliverfunk) [inline bluemira/equilibria/optimisation/problem/_nested_position.py:99]: I'm not sure if it's better to do it like this, it's more clear potentially to see the opt_conditions as they were. Not sure how nb that is though.

File: eudemo/eudemo/pf_coils/designer.py [hunk 3]
```diff
@@ -260,23 +248,28 @@ def _make_pulsed_coilset_opt_problem(
...
+                opt_conditions=eq_settings["optimisation_settings"]["conditions"],
+                gamma=eq_settings["gamma"],
+                relaxation=eq_settings["relaxation"],
+                peak_PF_current_factor=eq_settings["peak_PF_current_factor"],
+            ),
+            position_settings=PositionSettings(
+                problem=PulsedNestedPositionCOP,
+                algorithm=pos_settings["optimisation_settings"]["algorithm_name"],
+                opt_conditions=pos_settings["optimisation_settings"]["conditions"],
+            ),
```
Comment (oliverfunk) [inline eudemo/eudemo/pf_coils/designer.py:272]: Wow so much cleaner.

File: bluemira/equilibria/run.py [hunk 4]
```diff
@@ -81,48 +85,76 @@ class Snapshot:
...
 
-    eq: Equilibrium
+    eq: MHDState
     coilset: CoilSet
     constraints: Optional[CoilsetOptimisationProblem] = None
     profiles: Optional[Profile] = None
-    optimiser: Optional[EquilibriumOptimiser] = None  # noqa: F821
+    optimisation_result: [CoilsetOptimiserResult] = None
     limiter: Optional[Limiter] = None
     tfcoil: Optional[Coordinates] = None  # noqa: F821
```
Comment (oliverfunk) [inline bluemira/equilibria/run.py:102]: Coordinates not defined

File: bluemira/equilibria/optimisation/problem/_minimal_current.py (hunk 1 above)
Comment (je-cook) [reply bluemira/equilibria/optimisation/problem/_minimal_current.py:52]: sure

File: bluemira/equilibria/optimisation/problem/_nested_position.py (hunk 2 above)
Comment (je-cook) [reply bluemira/equilibria/optimisation/problem/_nested_position.py:99]: Its the mutable defaults again not for any readability improvements. I could possible put a debug statement with the contents of the opt_conditions like on L140 of wall silhouette
` ``python
       bluemira_debug(
            f"Setting up design problem with:\n"
            f"algorithm_name: {self.algorithm_name}\n"
            f"n_variables: {parameterisation.variables.n_free_variables}\n"
            f"opt_conditions: {self.opt_conditions}\n"
            f"opt_parameters: {self.opt_parameters}"
        )
```

Comment (je-cook) [reply bluemira/equilibria/optimisation/problem/_nested_position.py:99]: https://beta.ruff.rs/docs/rules/mutable-argument-default/

Comment (oliverfunk) [reply bluemira/equilibria/optimisation/problem/_nested_position.py:99]: I see yea.. Adding the defaults to the doc string would be fine as well

File: bluemira/equilibria/run.py [hunk 5]
```diff
@@ -218,73 +253,146 @@ def __init__(
...
+        else:
+            self.__bd_settings = BreakdownCOPSettings(**value)
+
+    @property
+    def eq_settings(self) -> EQSettings:
+        """Equilibrium COP settings."""
+        return self.__eq_settings
 
     @property
     def _eq_settings(self) -> EQSettings:
```
Comment (oliverfunk) [inline bluemira/equilibria/run.py:374]: Will this get dep'd infavour of just `eq_settings`?

Comment (je-cook) [reply bluemira/equilibria/run.py:374]: yep its already removed in #2320

File: bluemira/equilibria/optimisation/problem/_nested_position.py (hunk 2 above)
Comment (je-cook) [reply bluemira/equilibria/optimisation/problem/_nested_position.py:99]: sure I'll do that

File: bluemira/equilibria/run.py [hunk 6]
```diff
@@ -81,48 +86,76 @@ class Snapshot:
...
         The PF coil placement boundary
     """
 
-    eq: Equilibrium
+    eq: MHDState
     coilset: CoilSet
     constraints: Optional[CoilsetOptimisationProblem] = None
     profiles: Optional[Profile] = None
-    optimiser: Optional[EquilibriumOptimiser] = None  # noqa: F821
+    optimisation_result: Optional[CoilsetOptimiserResult] = None
```
Comment (oliverfunk) [inline bluemira/equilibria/run.py:101]: Missed the Optional here

File: bluemira/equilibria/run.py
Comment (oliverfunk) [inline bluemira/equilibria/run.py:1]: So much cleaner, looking forward to the dep removals.

File: bluemira/equilibria/run.py (hunk 6 above)
Comment (je-cook) [reply bluemira/equilibria/run.py:101]: already done :)

**Review by oliverfunk — COMMENTED** [review]
Mechanically it's looking really good.

My only critism would be I think all the COP's should have a more standard interface (with the typings and the docs for defaults etc.)

Will go over it again later, if you want to add any changes, before I approve.

File: bluemira/equilibria/optimisation/problem/_nested_position.py (hunk 2 above)
Comment (je-cook) [reply bluemira/equilibria/optimisation/problem/_nested_position.py:101]: Ive done one better see updates

File: bluemira/equilibria/run.py
Comment (je-cook) [reply bluemira/equilibria/run.py:1]: me too

File: eudemo/eudemo/comp_managers.py [hunk 7]
```diff
@@ -165,11 +167,20 @@ def _add_plugs(
         )
 
 
-class PortManagerMixin(OrphanerMixin):
+class PortManagerMixin(OrphanerMixin, abc.ABC):
```
[the remark this answers is not in the document]
Comment (je-cook) [reply eudemo/eudemo/comp_managers.py:170]: Discussed separately we may do some rearranging based on further abstraction

**Pull request merged**